### ISSUE TITLE
GPU acceleration: how much?

### ISSUE THREAD

**Issue by NetBUG** [alexjc/neural-doodle#51]
Dear @alexjc, what is expected acceleration when running on GPU?
I have two machines: 
1) MBP @ 2.2 GHz Intel Core i7 / 16 GB RAM / 1.5 Gb Intel Iris Pro
2) i7-5930K CPU @ 3.50GHz / 64 GB RAM / NVIDIA Titan X
I expected the second configuration to work significantly faster with GPU acceleration, but it's nearly the same on similar image size (23 minutes on GPU vs 42 minutes on Mac).

The CUDA config is `Using gpu device 0: GeForce GTX TITAN X (CNMeM is enabled with initial size: 98.0% of memory, CuDNN 4007)`
Is that difference expected or I have something misconfigured on the NVidia machine?
I have absolutely no experience with GPUs.

**Comment by alexjc** [comment]
The second machine should be much faster, a few minutes at most for the samples with the project. I don't know if it's misconfigured, the log line printed looks correct.

Can you run `nvida-smi dmon` at the same time as the program and see what the columns say?

**Comment by NetBUG** [comment]
Got it, seems I have 0% GPU load. 
Thanks, will research the problem now.
![image](https://cloud.githubusercontent.com/assets/1539823/13764385/eaacc926-ea0b-11e5-9b02-12a4ecb6624f.png)

**Comment by alexjc** [comment]
If you leave `dmon` on long enough, does it occasionally spike?

**Comment by NetBUG** [comment]
I don't understand exactly what is happening. 
`ourzhumt@sujbiswa-70-90:~/neural-doodle-master$ THEANO_FLAGS='floatX=float64,device=gpu0,lib.cnmem=1' python3 doodle.py --device=cpu --style samples/Renoir.jpg --output samples/Suriargb1.png --iterations=40
Using gpu device 0: GeForce GTX TITAN X (CNMeM is enabled with initial size: 98.0% of memory, CuDNN 4007)
Neural Doodle for semantic style transfer.
  - Using device `gpu0` for processing the images.`
However, gpu3 is heating (34 -> 47 degrees) and has non-idle frequencies. When I stop the script, it cools down and all 4 core have same 405/135 MHz clocks.
However, it still has zeros in other columns.

gpu0 had 2% occasionally.
![image](https://cloud.githubusercontent.com/assets/1539823/13764539/91e661e2-ea0d-11e5-8cf4-179782c3f3e6.png)

However, it looks like Theano is falling back to CPU, don't know why.
`>>> import theano
Using gpu device 1: GeForce GTX TITAN X (CNMeM is enabled with initial size: 98.0% of memory, CuDNN 4007)`

**Comment by alexjc** [comment]
If you use `--device` the script will set `THEANO_FLAGS` for you. Here you use `--device=cpu` but specify `THEANO_FLAGS` as using `gpu0`. I don't think that's the problem necessarily.

Also, try `float32` as it's trained that way. (Not sure if this will help.)

**Comment by carlthome** [comment]
I also have suspiciously low GPU usage reported with a GTX 980 (around 0-10% load throughout a semantic style transfer with the sample images).

```shell
python3 doodle.py --style samples/Gogh.jpg --content samples/Seth.jpg --output output.jpg --device=gpu0 --iterations=80
```

```shell
[global]
floatX = float16
device = gpu
exception_verbosity=high
optimizer=fast_compile
mode=FAST_RUN

[scan]
allow_gc = True

[nvcc]
fastmath = True

[lib]
cnmem = 0.1

[blas]
ldflags = -llapack -lblas
```

**Comment by alexjc** [comment]
Try removing the `fast_compile`, removing `exception_verbosity` and using `float32`.

**Comment by carlthome** [comment]
Doesn't seem to make a difference. The GPU utilization still hovers around
1%.

**Comment by carlthome** [comment]
Doesn't seem to make a difference. The GPU utilization still hovers around 1%.

**Comment by NetBUG** [comment]
@alexjc: I've tried to benchmark my system with convnet-benchmarks and realized that tests written for Pylearn and Theano do not perform as they should on GPU.
Seems it's either compilation problem or somehow else Theano is initializing GPU but not using it.
It's not your fault at all. If I discover how to fix it (provide all CUDA-related paths at compile time for Cython/Theano etc), I'll tell.
Java-based examples run perfectly on my system (about 3 times faster than provided in benchmark reference).

**Comment by NetBUG** [comment]
@carlthome @alexjc 
I've rebuilt everything (Numpy, Scipy, Theano) with CUDA/bin in $PATH and $CUDA_ROOT set.
Now full processing with float64 takes about 90 seconds. GPU is loaded up to 30-40% now, GPU memory is not largely occupied (I use small images).
I tried to recompile Theano only, that did not give any result.

Thus, one has to build these Python modules with not only nvidia software installed but also with manually set CUDA_ROOT.
Hope that helps.

**Comment by alexjc** [comment]
@NetBUG Glad to hear! 30%-40% sounds correct for the sample images. If memory usage is low enough you can run the script multiple times in parallel.

How were these libraries installed before, with `pip`?

**Comment by NetBUG** [comment]
Both times with `pip` but when I've built them initially, `$CUDA_ROOT` was not set. Seems they do not discover hardware changes dynamically.

**Comment by carlthome** [comment]
Surprising find, since I'm already successfully using the GPU with Theano in my current Python environment (the same that doesn't seem to work with neural-doodle). This following example from Theano's site, for example, works on my install:

```python
from theano import function, config, shared, sandbox
import theano.tensor as T
import numpy
import time

vlen = 10 * 30 * 768  # 10 x #cores x # threads per core
iters = 1000

rng = numpy.random.RandomState(22)
x = shared(numpy.asarray(rng.rand(vlen), config.floatX))
f = function([], T.exp(x))
print(f.maker.fgraph.toposort())
t0 = time.time()
for i in range(iters):
    r = f()
t1 = time.time()
print("Looping %d times took %f seconds" % (iters, t1 - t0))
print("Result is %s" % (r,))
if numpy.any([isinstance(x.op, T.Elemwise) for x in f.maker.fgraph.toposort()]):
    print('Used the cpu')
else:
    print('Used the gpu')
```

So I don't see why a reinstall should be necessary, assuming neural-doodle's GPU support is Theano-based. Which Theano version did you build? I'm using 0.8.0rc1. Cause for trouble?

**Comment by NetBUG** [comment]
@carlthome I used same version.
Unfortunately old configuration was not preserved while reinstalling.
However the symptoms looked exactly the same: when I did `import theano`, I got a line `Using gpu device 0: GeForce GTX TITAN X (CNMeM is enabled with initial size: 98.0% of memory, CuDNN 4007)` but in fact no GPU or video memory was used. As far as I understand, all three modules (numpy, scipy and theano) use native extensions which are compiled when you install them via pip. 
Seems last time some of them has not detected CUDA libraries and did not include support/bindings for it without `$CUDA_PATH`. This is not a professional answer, it's my first time working on a computer with GPU ;)
P.S. Your code only uses GPU if THEANO_FLAGS are set for it.
